### PR TITLE
Autobuild: Clean up workflow description

### PR DESCRIPTION
**Pull request by hoffie**
This is a small, split-out part of the announced autobuild refactoring.

**Short description of changes**

- Autobuild: Drop commented out code and duplicate newlines
- Autobuild: Shorten branch [matching syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet)
- Autobuild: Drop comments which repeat step names
- Autobuild: Unify identation
- Autobuild: Reorganize, shorten and strip redundant comments
- Autobuild: Shorten if expressions
  [`${{ ... }}` is optional for if: entries](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idif). Therefore drop it in order to reduce clutter.
- Autobuild: Line-wrap long if statements
- Autobuild: Rename uses_codeql -> run_codeql

CHANGELOG: Autobuild: Refactor for better readability

**Context: Fixes an issue?**

Improve readability/maintenance.

**Does this change need documentation? What needs to be documented and how?**

No.

**Status of this Pull Request**

Ready when CI is green.

**What is missing until this pull request can be merged?**

Ready when CI is green.

Note: I suggest reading individual commit diffs if unsure. `git show -w` and Github's whitespace-ignoring diff may also help during review.

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:48]:
```suggestion
         - name:                    Get Jamulus build/release info, determine actions & variables
```

Although not part of this PR, I don't like this description at all. What is "determine actions"?

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:306]: Maybe leave this link here and delete it below.

File: .github/workflows/autobuild.yml
Comment (hoffie) [inline .github/workflows/autobuild.yml:207]: What about this:

```suggestion
      - name:                       "Install build dependencies for ${{ matrix.config.config_name }}"
```

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:306]: Hm, I'd rather like to remove it. Reasoning:
- While this might have been a new feature at some point, it no longer is. It's standard Github Workflow syntax and not some kind of hard to find trick or something. Anyone who touches the workflow should read the documentation. I'm considering this to be a rather foundational thing and we don't link foundational things in other places either (e.g. C++ or Qt books in source files ;)).
- I don't know why it's worth to document this feature (the "outputs" logic) right here. The whole workflow is sprinkled with `outputs` references as it's such a basic idiom. So if we document it at all, it should be documented at the very top of the file. Maybe I'm missing something, but `grep -cF '.outputs.' .github/workflows/autobuild.yml` returns 33 places where outputs are either set or read. I don't see why those two occurences would be special.

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:48]:
```suggestion
         - name:                    Determine release version, type and prerelease variables
```

At least, that's what the script does. I'm open to other wordings though. :)

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:306]: > Of course the "can also be warn,…" can be removed but the links to the documentation or blog posts should remain.

(I've replied to that directly in the other thread)

> Some comments might be trivial (Checkout code,…)

I'm a huge fan of comments to explain specialities. I'm annoyed by comments which state the obvious. They just make the file longer and make me think "umm, I've just read that already". It's a [code smell](https://moderatemisbehaviour.github.io/clean-code-smells-and-heuristics/comments/c3-redundant-comment.html).
You've already mentioned one extreme case. Here's another.
https://github.com/jamulussoftware/jamulus/blob/master/.github/workflows/autobuild.yml#L238-L239

Thanks for the specific comments, that's much easier to work with. :)

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:48]: Yes. That's better.

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:207]: Yes. I think that's better (which would also imply a change of all the script file names...)

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:306]: Ok. Probably I still think of the old times where the .yaml was basically only a three step procedure (and where I wanted it to be clarified by nefarius). 

So I'll need to be happy with it.

Comment (hoffie) [reply .github/workflows/autobuild.yml:306]: > So I'll need to be happy with it.

Well, I don't want to say "it must be like that". I want to outline what my goals are. If I'm missing something or if there are better approaches, I'm open to that. But it should have some justification :)

Comment (ann0see) [reply .github/workflows/autobuild.yml:306]: The thing is  - in my opinion - the whole autobuild file has become complicated (there are many specialities. Let me do another review on how I would comment the code)

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:1]:
```suggestion
```

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:3]: Maybe a brief comment like

"This file builds the source and produces artifacts for every operating system. If requested it creates a release and uploads the binaries to GitHub releases. Moreover, it tests the code with CodeQl"

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:28]:
```suggestion
      # Only automatically run the build on pushes to master to avoid wasting ressources
```

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:31]: Below at create_release:

Check if we are doing a release or just a normal build. This must be done before actually building the app to clarify where to upload the binaries and if we need to create a release.

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:58]: I think this was mainly a bug-fix or related to pushing a release tag? Not sure why this was needed again (ie. why we needed to check for an existing release. But it's good to have it anyways.)

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:76]: Start the actual build process.

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:77]:
```suggestion
    name:                           Build for ${{ matrix.config.config_name }}
```

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:174]: Set up build environment. This downloads dependencies if needed or uses those cached from previous runs.

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:234]: Should probably be named something like

Move files... But that's not part of this PR though.

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:260]: Upload/Release binaries

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:266]: Why is this done after the upload artifact call? Couldn't there also be signed non-release artifacts?

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:298]: If requested upload artifacts to release.

**Review by ann0see — COMMENTED** [review]
I think I'm probably commenting a lot about the structure? But please see below:

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:266]: Hm, currently it only runs on release builds. It could probably also be run earlier. Not sure if there are any reasons for not signing dev builds, but as this runs on @emlynmac's fork and not on ours anyway, I'm not sure how deeply we should care...? :)

Maybe open a separate issue for that?

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:58]: Might also be relevant if we encounter any issues with the PR run and have to re-start it. I'd keep it (especially as part of this PR).

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:31]: Hm, I'd prefer improving the name instead. What about this:

```diff
- name:                          Prepare Auto-Build/Release
+ name:                          Create Github release if this is a release build
```

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:76]: Hm, what would be the added value of such a comment? It's just a more elaborate re-phrasing of the `name:` below to me.

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:77]: Github-wise the term is `assets`, so if we want to provide details, we should probably keep this as-is?

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:234]: Yes, that could indeed be improved to match what's actually done in the step (in another PR).

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:174]: This sounds like some kind of headline to me, but technically there is no such thing. All of that is done as part of the `release_assets` job. It doesn't feel right to add such a "block comment" in a place where there is no block.

What are you mainly after at? Maybe adding the "Download dependencies" part to the rather inaccurate "Prepare" step later would be beneficial? I'll add a suggestion below.

I'm all for clarifying the structure of this file. Github's workflow syntax already provides us with a way to technically annotate each step (`name:`). I think its use is even mandatory. Therefore, I think we should do all the useful optimizations there and don't add another level of structure via comments.
Improving `name:` has the benefits of also improving the CI logs in the same go. A `grep '[ -] name:' .github/workflows/autobuild.yml | grep -v artifact` can also be used to get a sense of the structure of this workflow.

File: .github/workflows/autobuild.yml
Comment (hoffie) [inline .github/workflows/autobuild.yml:207]:
```suggestion
      - name:                       "Download and install build dependencies for ${{ matrix.config.config_name }}"
```

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:260]: The step in question actually just uploads the artifact to the Actions job. I think the `name:` already covers that. The upload to the release does not happen here, but in a later step.

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:30]: Background: https://github.com/jamulussoftware/jamulus/pull/978#issuecomment-778540855

So... now I understand what was probably meant by the original comment. But rather than explaining Why (CodeQL mandates that) it explains the What (the fact that this list is a subset of the other can be derived without the comment). I've now added a comment to explain the reasoning. I've also added the above suggestion.

File: .github/workflows/autobuild.yml
Comment (softins) [reply .github/workflows/autobuild.yml:33]: Looking at the workflow, I think this name is slightly misleading. This step happens first, before the builds, so it isn't actually creating the Github release, just setting up variables to tell later steps to generate one. So "Prepare" or "Setup" is better, but happy for the rest of the name to be made clearer.

File: .github/workflows/autobuild.yml
Comment (hoffie) [inline .github/workflows/autobuild.yml:38]: > Looking at the workflow, I think the new name is slightly misleading. This step happens first, before the builds, so it isn't actually creating the Github release, just setting up variables to tell later steps to generate one. So "Prepare" or "Setup" is better, but happy for the rest of the name to be made clearer.

@softins By *new name* you are referring to the line which this comment is attached to, right?

This job does indeed create the Github release (although it doesn't fill it). The step some lines below is also named like that.

Maybe this would be better:
```suggestion
    name:                           Create empty Github release if this is a release build
```

Comment (hoffie) [reply .github/workflows/autobuild.yml:38]: Addition: I just realized that the changelog is generated by `anaylze_git_references.py`. Generating a file in an "analyze_" script certainly violates the principle of least surprise...

As this is both related to this PR and #2471, I think I'll open a separate Issue to track this and work on it when both PRs are merged. I've adjust the name of the `analyse_git_reference.py` step for now.

Comment (softins) [reply .github/workflows/autobuild.yml:38]: > > Looking at the workflow, I think the new name is slightly misleading. This step happens first, before the builds, so it isn't actually creating the Github release, just setting up variables to tell later steps to generate one. So "Prepare" or "Setup" is better, but happy for the rest of the name to be made clearer.
> 
> @softins By _new name_ you are referring to the line which this comment is attached to, right?
> 
> This job does indeed create the Github release (although without the artifacts, of course). The step some lines below is also named like that.
> 
> Example: https://github.com/jamulussoftware/jamulus/runs/5265414702?check_suite_focus=true#step:5:1

Ah, ok, I misunderstood. I thought it was just preparing variables for later use. So it's actually creating a release container that will later receive the assets.

My only other comment there would be that the `name` is getting rather long, and becoming more like a comment or description. Can it be made more concise but still accurate?

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:30]: Yes. That's good.

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:65]: I think we could even drop the variables here (most times they're not even filled).

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:80]: Yes, but I'm unsure if it adds any value.

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:177]: > Improving name: has the benefits of also improving the CI logs

Good point. And yes 'Prepare' sounds impercise.

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:210]: +1

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:269]: @emlynmac could you please comment why this is here and not before any upload has started?

Comment (emlynmac) [reply .github/workflows/autobuild.yml:269]: This step is to notarize the build, which has to happen after the build is done and the artifacts created.
The signing of the macOS build is a two-step process; build with certs then notarize the full thing with Apple.

Comment (ann0see) [reply .github/workflows/autobuild.yml:270]: Is this basically a step which transfers something (hashes?) to apple?

**Comment by ann0see** [comment]
BTW: See the nightly: https://github.com/jamulussoftware/jamulus/runs/5487103016?check_suite_focus=true

The Create Release step doesn't even output RELEASE_TITLE

File: .github/workflows/autobuild.yml
Comment (ann0see) [inline .github/workflows/autobuild.yml:66]: I think a $ is missing:

```suggestion
      - name:                       'Create Release ${{steps.jamulus-build-vars.outputs.RELEASE_TAG}}  ${{steps.jamulus-build-vars.outputs.RELEASE_TITLE}}'
```

**Review by ann0see — CHANGES_REQUESTED** [review]

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml]: Great. Thanks.

File: .github/workflows/autobuild.yml.orig
Comment (ann0see) [inline .github/workflows/autobuild.yml.orig:1]: What is this file? Probably a backup?

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml]: > My only other comment there would be that the name is getting rather long, and becoming more like a comment or description. Can it be made more concise but still accurate?

Hrm, you are right. I struggle to find good wording though. Maybe we should drop the imperative for job-level names.

If just adjusted it like that:
```suggestion
    name:                           Build vars & Github release (if required)
```

File: .github/workflows/autobuild.yml.orig
Comment (hoffie) [reply .github/workflows/autobuild.yml.orig]: Argh. Yep. Removing.

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml]: Ok, removed. It also matches the first job's name better, now that the imperative has been dropped.

**Review by ann0see — APPROVED** [review]

**Comment by ann0see** [comment]
I think I‘m happy now v

**Comment by hoffie** [comment]
Currently still cleaning up the git history. As @softins had some comments as well I'll wait with merging until tomorrow (or explicit feedback).

File: .github/workflows/autobuild.yml
Comment (emlynmac) [reply .github/workflows/autobuild.yml]: There's a bunch of information about how notarization works here: https://developer.apple.com/documentation/security/notarizing_macos_software_before_distribution

File: .github/workflows/autobuild.yml
Comment (softins) [inline .github/workflows/autobuild.yml:209]: Do we need to say "Download and"? Downloading is probably implicit.

File: .github/workflows/autobuild.yml
Comment (softins) [inline .github/workflows/autobuild.yml:285]: Not sure the "if this is a release" is needed. If it's not a release, the step will be shown in the log as not executed anyway. Same for artifact 2 below.

**Review by softins — APPROVED** [review]
Looking good. I left a couple of comments regarding step naming, but they are only suggestions, not deal-breakers :)

Approving anyway.

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:209]: I had added that as a response to feedback by @ann0see.
@ann0see Should we keep "Download and" in or should I shorten it again?

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml:285]: I might have added this in response to feedback as well, but I've lost track by now...
@ann0see Should I shorten it as well? :)

Comment (ann0see) [reply .github/workflows/autobuild.yml:285]: Probably it’s redundant so yes, you can remove it.

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml:209]: Maybe use Set up? But that would be close to the unwanted „Prepare“ again.

Comment (hoffie) [reply .github/workflows/autobuild.yml]: Ok, done.

File: .github/workflows/autobuild.yml
Comment (hoffie) [reply .github/workflows/autobuild.yml]: Done. Will merge on CI green.

File: .github/workflows/autobuild.yml
Comment (ann0see) [reply .github/workflows/autobuild.yml]: @hoffie sorry for the small modification request: Could you please add a comment here that this needs to be done after the build is created?

Comment (hoffie) [reply .github/workflows/autobuild.yml]: Not sure I understand.
Do you mean that the binaries can only be signed once they're built? That would be rather obvious to me (similar for other instances: We don't explicitly highlight that the build dependencies have to be prepared before the build can start).

Or do you mean that the binaries can only be signed once the artifact has been uploaded to the job? I don't think this is the case (or if it is, I don't understand why). The Notarize step references `product-path:             deploy/${{ steps.step_cmd3_postbuild.outputs.artifact_1 }}`. This looks like a filesystem path to me and not like a job artifact path. So I guess the order could be turned around. Adding a comment which states the contrary would seem strange then :)

Maybe you could send a follow-up PR with a suggested wording?

(In general, I think a PR which touches specific code should leave the code in better state than it was before. However, we should be cautious to avoid trying to push to much *additional* things which *could* or *should* be improved into PRs. PRs are getting large and take lots of iterations until they can get merged...)

Comment (ann0see) [reply .github/workflows/autobuild.yml]: Yes sure. @emlynmac should open a PR then since he knows what to do. I think we can move it around - at least from my readings.

**Comment by ann0see** [comment]
Ok. Ready to merge after Android build is finished.

**Comment by ann0see** [comment]
@hoffie please merge (squash merge?) this if it's finished.

**Comment by hoffie** [comment]
> squash merge?

Although the PR contains more than a few commits, I'd like to keep it separate to keep the information what was changed for what reasons. I've already squashed all the same-type adjustments (wording improvements are a single commit, for example).

**Pull request merged**
